File: Research/auto-regresive-tranformer/Production-Code/auto-regressive-tf.cpp
#include <torch/torch.h>
#include <torch/cuda.h>
#include <vector>
#include <iostream>
#include "autoreg_transformer.h"

int main() {
    // Define hyperparameters
    int num_layers = 6;
    int input_size = 100;
    int hidden_size = 512;
    int output_size = 100;
    float learning_rate = 0.0005;
    int batch_size = 64;
    int num_epochs = 10;

    // Load training data
    std::vector<std::vector<int>> input_seqs, output_seqs;
    // Load training data...

    // Initialize autoregressive transformer model
    AutoRegTransformer model(num_layers, input_size, hidden_size, output_size);

    // Initialize optimizer
    torch::optim::Adam optimizer(model.parameters(), torch::optim::AdamOptions(learning_rate));

    // Enable mixed precision training
    model.to(torch::kHalf);
    optimizer = torch::optim::Adam(model.parameters(), torch::optim::AdamOptions(learning_rate).set_dtype(torch::kHalf));

    // Train model for multiple epochs
    for (int epoch = 1; epoch <= num_epochs; epoch++) {
        // Iterate over training data
        float epoch_loss = train(model, input_seqs, output_seqs, optimizer, batch_size);

        std::cout << "Epoch " << epoch << " loss: " << epoch_loss << std::endl;
    }

    return 0;
}

File: Research/Auto-regressive-transformer/auto-regressive-tf.py
import tensorflow as tf

# Define autoregressive transformer model
class AutoRegTransformer(tf.keras.Model):
    def __init__(self, num_layers, input_size, hidden_size, output_size):
        super(AutoRegTransformer, self).__init__()
        self.num_layers = num_layers
        self.hidden_size = hidden_size

        # Define input embedding layer
        self.embedding = tf.keras.layers.Embedding(input_size, hidden_size)

        # Define transformer encoder layers
        self.encoder_layers = [tf.keras.layers.TransformerEncoderLayer(hidden_size, num_heads=8,
                                    ff_dim=2048) for _ in range(num_layers)]
        self.encoder = tf.keras.layers.TransformerEncoder(self.encoder_layers, num_layers=num_layers)

        # Define output linear layer
        self.linear = tf.keras.layers.Dense(output_size)

    def call(self, input_seq, training=True):
        # Embed input sequence
        embedded_seq = self.embedding(input_seq)

        # Encode embedded sequence using transformer encoder
        encoded_seq = self.encoder(embedded_seq, training=training)

        # Apply linear layer to output sequence
        output_seq = self.linear(encoded_seq)

        return output_seq

# Define training loop function
@tf.function
def train(model, data_loader, optimizer, grad_clip_norm=1.0, dropout_rate=0.1):
    # Iterate over training data
    total_loss = 0.0
    num_time_steps = 0
    for input_seq, output_seq in data_loader:
        # Feed input sequence through model to generate output sequence
        with tf.GradientTape() as tape:
            predicted_output_seq = model(input_seq, training=True)
            # Calculate cross-entropy loss for each time step
            loss = tf.reduce_mean(tf.keras.losses.sparse_categorical_crossentropy(
                                output_seq, predicted_output_seq, from_logits=True))

        # Sum up losses for sequence and update total loss and number of time steps
        total_loss += loss * tf.shape(input_seq)[0]
        num_time_steps += tf.shape(input_seq)[0]

        # Calculate gradients and clip by norm
        gradients = tape.gradient(loss, model.trainable_variables)
        gradients, _ = tf.clip_by_global_norm(gradients, grad_clip_norm)

        # Add weight decay to gradients
        for var, grad in zip(model.trainable_variables, gradients):
            grad += 0.001 * var

        # Apply dropout to gradients
        gradients = [tf.nn.dropout(grad, dropout_rate) for grad in gradients]

        # Backpropagate and update model parameters
        optimizer.apply_gradients(zip(gradients, model.trainable_variables))

    # Average losses over all sequences in batch to get epoch loss
    epoch_loss = total_loss / num_time_steps

    return epoch_loss

# Define hyperparameters

# 456 layers of hidden and output
num_layers = 6
hidden_size = 512
learning_rate = 0.0005
batch_size = 64
num_epochs = 10

# Load training data
training_data = tf.data.Dataset.from_tensor_slices((input_seqs, output_seqs))
training_data = training
